### PR TITLE
Refs # Handle more than 2 events

### PR DESCRIPTION
**Pull request by saimekala07**
Adding eligible criteria to group more than 2 events.
Adding additional action to handle retro events

### CODE REVIEW

File: app/models/enrollment_action/base.rb [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
     attr_reader :action
+    attr_reader :additional_action
     attr_reader :termination
     attr_reader :errors
 
-    def initialize(term, init)
+    def initialize(term, init, additional_init = nil)
       @termination = term
       @action = init
+      @additional_action = additional_init
```
Comment (saimekala07) [inline app/models/enrollment_action/base.rb:11]: any better name suggestions for `additional_action`?

File: app/models/enrollment_action/base.rb [hunk 2]
```diff
@@ -21,6 +23,7 @@ def update_business_process_history(entry)
 
     def self.select_action_for(chunk)
       selected_action = [
+        ::EnrollmentAction::RetroContinuityAndTerm,
```
Comment (saimekala07) [inline app/models/enrollment_action/base.rb:26]: any better name suggestions for enrollment actions

File: app/models/external_events/enrollment_event_notification.rb [hunk 3]
```diff
@@ -326,17 +326,41 @@ def kind
...
+    def coverage_year
+      if is_shop?
+        plan_year = find_employer_plan_year(policy_cv)
+        (plan_year.start_date..plan_year.end_date)
+      else
+        (Date.new(active_year.to_i, 1, 1)..Date.new(active_year.to_i, 12, 31))
+      end
+    end
+
     def is_adjacent_to?(other)
```
Comment (TreyE) [inline app/models/external_events/enrollment_event_notification.rb:338]: Couple issues here.  One of the most important parts here is knowing that was have impacted how chunks are constructed only for this specific case - and left it alone for everyone else.

Given the above:
- I don't see how we construct chunks in a way that considers 3 events while only considering 2 at a time.  We might need to go up a level and find a way to consider an arbitrary set of chunks, possibly by doing more than one 'splitting' pass, or very early keeping any chunks which net out to exactly 2 chunks.
- I don't see how we can be sure that it affects this case and only this case without a spec, and without executing some examples that show you get what you expect for both the old and new behaviour.  Most likely the spec will need to involve 'jumping up a level' to app/business_processes/handlers/enrollment_event_enrich_handler.rb - or mimicing it's behaviour in order to make sure chunks are organized as we expect without impacting existing functionality.

**Review by TreyE — CHANGES_REQUESTED** [review]
Requested changes, missing spec and I don't understand how we are proving we didn't change anything.

File: app/models/external_events/enrollment_event_notification.rb (hunk 3 above)
Comment (saimekala07) [reply app/models/external_events/enrollment_event_notification.rb:338]: Currently we are handling/checking an event adjacent is another event and grouping them, if all the event happened in a given year only.

Added new logic will check for event adjacent is another event across the year for the specific retro case.
Scenario1:
-  group 3 events
Scenario1:
The current year(ex: 2021) canceled not terminated
Added retro coverage(ex: 2020)
Renewal created.

- group 2 events
Scenario2:(selected optional not to generate renewal )
The current year(ex: 2021) canceled not terminated
Added retro coverage(ex: 2020).

- plan change and dependent drop/add's that grouped happening for a given year shouldn't have any impacts by the new changes.

**Review by TreyE — CHANGES_REQUESTED** [review]
I've re-worked and simplified how we identify 3 enrollments at a time that qualify for the special treatment, and leave the rest of them alone.

I think we need to take a quick look at the specs you wrote to make sure that either they aren't testing the wrong thing, or, more likely, they found a bug in my implementation.

This approach should allow us to keep the existing behaviour more or less how it is, while grabbing only the very special cases where 3 enrollments need to be considered.

**Review by TreyE — APPROVED** [review]
Looks correct.

**Pull request merged**